sympy/solvers/solvers.py

```diff
@@ -2653,22 +2654,45 @@ def _tsolve(eq, sym, **flags):
                     return _solve(lhs - rhs, sym, **flags)
             elif sym not in lhs.exp.free_symbols:
                 return _solve(lhs.base - rhs**(1/lhs.exp), sym, **flags)
-            elif not rhs and sym in lhs.exp.free_symbols:
-                # f(x)**g(x) only has solutions where f(x) == 0 and g(x) != 0 at
-                # the same place
-                sol_base = _solve(lhs.base, sym, **flags)
-                if not sol_base:
-                    return sol_base  # no solutions to remove so return now
-                return list(ordered(set(sol_base) - set(
-                    _solve(lhs.exp, sym, **flags))))
-            elif (rhs is not S.Zero and
-                        lhs.base.is_positive and
-                        lhs.exp.is_real):
-                return _solve(lhs.exp*log(lhs.base) - log(rhs), sym, **flags)
-            elif lhs.base == 0 and rhs == 1:
-                return _solve(lhs.exp, sym, **flags)
             else:
-                raise NotImplementedError
+                # sym in lhs.exp.free_symbols
+                if not rhs:
+                    # f(x)**g(x) only has solutions where f(x) == 0 and g(x) != 0 at
+                    # the same place
+                    sol_base = _solve(lhs.base, sym, **flags)
+                    if not sol_base:
+                        return sol_base  # no solutions to remove so return now
+                    return list(ordered(set(sol_base) - set(
+                    _solve(lhs.exp, sym, **flags))))
+                elif lhs.base.is_positive and lhs.exp.is_real:
+                    return _solve(lhs.exp*log(lhs.base) - log(rhs), sym, **flags)
+                elif rhs.is_Integer:
+                    if  lhs.base == 0:
+                        return _solve(lhs.exp, sym, **flags)
+                    else:
+                        e0=[]
+                        e1=[]
+                        sol=[]
+                        if rhs==1:
+                            e0 = _solve(lhs.exp, sym, **flags)
+                        x=divisors(rhs)
+                        for i in x:
+                            base_sols = solve(lhs.base - i, sym) + solve(lhs.base + i, sym)
+                            for s in set(base_sols):
+                                if eq.subs(sym, s) == 0:
+                                    sol.append(s)
+                        rewrite = lhs.rewrite(exp)
+                        if rewrite != lhs:
+                            try:
+                                e1= _solve(rewrite - rhs, sym, **flags)
+                            except:
+                                pass
+                        sol = sol + list(e1)+ list(e0)
+                        if len(sol)==0:
+                            raise NotImplementedError
+                        return list(set(ordered(sol)))
```

Comment:
You're calling `ordered` before `set`. It should be the other way round:
```python
return list(ordered(set(sol)))
```

Reply:
Maybe `list` isn't needed if `ordered` already returns a `list`...

Reply:
Ohh Previously I had written this ```return list(ordered(set(sol)))```  but  while commiting the last commit I accidently did undo. Okk now I had changed it. Thanks @oscarbenjamin

Reply:
list is just to make more robust.